feat(router): Set a different browser URL from the one for route matching

atscott opened the pull request:
This feature adds a property to the `NavigationBehaviorOptions` that allows developers to define a different path for the browser's address bar than the one used to match routes. This is useful for redirects where you want to keep the browser bar the same as the original attempted navigation but redirect to a different page, such as a 404 or error page.

fixes #17004

jnizet commented:
I had the feeling that the preferred way of redirecting was to return a UrlTree from the guard rather than imperatively navigating from the guard. 
How would we get this behavior when returning a UrlTree?

atscott commented:
> I had the feeling that the preferred way of redirecting was to return a UrlTree from the guard rather than imperatively navigating from the guard.

Yes, redirects should generally go through that mechanism. `router.navigate...; return false` isn't a true redirect as far as the Router's concerned. It's a new navigation that cancels and supersedes the current one. Redirects with the `UrlTree` return allow keeping the original `Promise` from `navigate` and not resolving it until the redirect navigation completes.

> How would we get this behavior when returning a UrlTree?

Unfortunately the `UrlTree` redirects don't provide a way to defined `extras` and was a design decision at the time (https://github.com/angular/angular/issues/27148#issuecomment-442265173). I think it's clear now that this wasn't right and it'll be available in v18 with https://github.com/angular/angular/pull/45023 (expanding the type is a breaking change).

It should and will be available, but I didn't feel that this feature needed to be blocked on having that available though.

> Also, I must say that using UrlTrees is quite confusing, especially for beginners. Common navigations (using routerLink or router.navigate) use strings or arrays of segments to navigate, but in order to redirect, we have to transform the string/array into a UrlTree, which is not intuitive.

Agreed. There are at least two reasons this PR is still in draft mode and one of them is that I'm not yet decided on the behavior here with `urlHandlingStrategy` and `urlSerializer` (that includes `string` as an option). At the moment, I'm leaning towards not serializing a string and not using `UrlHandlingStrategy.merge` with the property value. In this way, the `browserUrl` property can also be used as a way to manually avoid `UrlHandlingStrategy` behavior.

> Even with the current state of this PR, if the goal is to keep the currently requested URL, couldn't we simply pass a boolean option rather than an UrlTree option?

Two things here: It is _a_ goal but I wouldn't say it's necessarily the _only_ goal. The feature as proposed here is a lot more flexible than a `boolean` and could be used for other use-cases. Additionally, as described above, the `router.navigate...; return false;` pattern isn't really a redirect. I don't think a `boolean` directly in the navigation extras that depends on the current router/navigation state (if there is one) is a great API. That said, I do think we could consider a shortcut to getting this behavior with a `boolean` when `RedirectCommand` is available since that, by definition, is only working in the context of an actual redirect.

jnizet commented:
Thanks for the detailed explanations @atscott

atscott committed:
feat(router): Set a different browser URL from the one for route matching

This feature adds a property to the `NavigationBehaviorOptions` that
allows developers to define a different path for the browser's address
bar than the one used to match routes. This is useful for redirects
where you want to keep the browser bar the same as the original
attempted navigation but redirect to a different page, such as a 404 or
error page.

fixes #17004

ChristofFritz commented:
Thank you so much for building this @atscott ! We'll be using this the second it get's released.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

thePunderWoman reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented:
This PR was merged into the repository by commit 1d3a7529b4fa3617a5d6a97e742cb13818253a14.

The changes were merged into the following branches: main
